source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1060 @@
+===================
+Unified Test Format
+===================
+
+:Spec Title: Unified Test Format
+:Spec Version: 1.0
+:Author: Jeremy Mikola
+:Advisors: Prashant Mital
+:Status: Draft
+:Type: Standards
+:Minimum Server Version: N/A
+:Last Modified: 2020-08-15
+
+.. contents::
+
+--------
+
+Abstract
+========
+
+This project defines a unified schema for YAML and JSON specification tests,
+which run operations against a MongoDB deployment. By conforming various spec
+tests to a single schema, drivers can implement a single test runner to execute
+acceptance tests for multiple specifications, thereby reducing maintenance of
+existing specs and implementation time for new specifications.
+
+META
+====
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`_.
+
+Specification
+=============
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as a document by the test runner. This section defines the top-level
+keys for that document and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+The test format is also defined in the accompanying `schema.json <schema.json>`_
+file, which may be used to programmatically validate YAML and JSON files using
+a tool such as `Ajv <https://ajv.js.org/>`_.
+
+The top-level fields of a test file are as follows:
+
```

Comment:
I know this isn't ready for review yet, but I think one key thing the format is currently missing is a `testSchemaVersion` field. This field would essentially record which version of the unified test format spec the file conforms to thereby allowing runners to ensure that they are not running a test incorrectly (e.g. a field is added to each entry in `operations` but the runner is written for the previous format version and hence silently ignores this newly added field). The motivation for this field came from the experience of rolling out the CRUD V2 format.

In addition to the field itself, I had planned to add a section to the spec detailing how the `testSchemaVersion` would need to be updated and how we expect drivers to migrate asynchronously to newer formats as they become available. `testSchemaVersion` would follow a semantic-versioning like scheme and the version would only be updated to 2.0 when a format change would render existing runners either incapable of running the tests or prone to running them erroneously.